File: week01/frontend/src/api/tags.ts
import { apiFetch } from './client'
import type { Tag } from '@/types'

export function fetchTags(withCount?: boolean): Promise<Tag[]> {
  const qs = withCount ? '?with_count=true' : ''
  return apiFetch<Tag[]>(`/tags${qs}`)
}

export function createTag(name: string): Promise<Tag> {
  return apiFetch<Tag>('/tags', {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
}

export function deleteTag(id: string): Promise<void> {
  return apiFetch<void>(`/tags/${id}`, { method: 'DELETE' })
}

File: week01/backend/app/routers/tags.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app import crud
from app.database import get_db
from app.schemas.tag import TagCreate, TagResponse

router = APIRouter(prefix="/tags", tags=["tags"])


@router.get("/", response_model=list[TagResponse])
def list_tags(
    with_count: bool = False,
    db: Session = Depends(get_db),
) -> list[TagResponse]:
    return crud.get_tags(db, with_count=with_count)


@router.post("/", response_model=TagResponse, status_code=status.HTTP_201_CREATED)
def create_tag(data: TagCreate, db: Session = Depends(get_db)) -> TagResponse:
    existing = crud.get_tag_by_name(db, data.name)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Tag name already exists",
        )
    return crud.create_tag(db, data)


@router.delete("/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_tag(tag_id: UUID, db: Session = Depends(get_db)) -> None:
    tag = crud.get_tag(db, tag_id)
    if not tag:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    crud.delete_tag(db, tag)
